detect/decoder.go

```diff
@@ -0,0 +1,150 @@
+package detect
+
+import (
+	"bytes"
+	"encoding/base64"
+	"regexp"
+	"unicode"
+)
+
+var b64Regexp = regexp.MustCompile(`[\w\/\-\+]{16,}={0,3}`)
+var decoders = []func(string) ([]byte, error){
+	base64.StdEncoding.DecodeString,
+	base64.RawURLEncoding.DecodeString,
+}
+
+// EncodedSegment represents a portion of text that is encoded some way.
+// Decoded values can also have encoded text in them so there can be a kind of
+// tree of segments needing to be decoded. This is why the term parent is used
+// in a few places below.
+type EncodedSegment struct {
+	// relative (rel) vs absolute (abs) values:
+	//
+	// If a value is double encoded, multiple passes happen do decode it all.
+	// Relative values are the bounds of the encoded value in the current pass.
+	// Absolute values refer to the bounds of the original encoded segment so
+	// that we can keep track of locations in the original data.
+	//
+	relStart     int
+	relEnd       int
+	absStart     int
+	absEnd       int
+	decodedValue string
+}
+
+// decode returns the data with the values decoded in-place
+func decode(data string, parentSegments []EncodedSegment) (string, []EncodedSegment) {
+	segments := findEncodedSegments(data, parentSegments)
+
+	if len(segments) > 0 {
+		result := bytes.NewBuffer(make([]byte, 0, len(data)))
+
+		relStart := 0
+		for _, segment := range segments {
+			result.WriteString(data[relStart:segment.relStart])
+			result.WriteString(segment.decodedValue)
+			relStart = segment.relEnd
+		}
+		result.WriteString(data[relStart:])
+
+		return result.String(), segments
+	}
+
+	return data, segments
+}
+
+// findEncodedSegments finds the encoded segments in the data and maps them to
+// any parent segments from a previous pass.
+func findEncodedSegments(data string, parentSegments []EncodedSegment) []EncodedSegment {
+	if len(data) == 0 {
+		return []EncodedSegment{}
+	}
+
+	matchIndices := b64Regexp.FindAllStringIndex(data, -1)
```

Comment:
Would it be paranoid to do a second pass and filter out any values that are all lowercase or uppercase? (Random thought.)

I suspect this would filter a large number of false positives without the overhead of calling base64 decode.

Reply:
Good note!

I'm planning to do some performance tests to see what I can do to bail out as early as possible with real world code, but I hadn't considered specifically that one.

My current list of ideas to try in different combinations to see if they help speed it up:

- have the `decodedMap` persist across the different passes so that only new things would be decoded
- entropy check (if it won't filter out real b64)
- (newly added) the ideas you just listed
- other misc checks that look at what's highly unlikely in things that aren't base64 (what those are are TBD and I'd only do that if the above didn't get us near the same performance as gitleaks without the changes).

Reply:
Generating some stats to figure out how likely different situations are.

`test-data` file:

```sh
git clone https://github.com/kubernetes/kubernetes.git
git -C kubernetes/ log --patch --all > test-data
```

Test script:

```python3
from collections import defaultdict
from secrets import randbits
from base64 import b64encode
from string import ascii_lowercase
from string import ascii_uppercase
from string import digits

lowercase = set(ascii_lowercase)
uppercase = set(ascii_uppercase)
digits = set(digits)
special = {'+', '/'}

count = defaultdict(int)
with open("test-data", "rb") as test_file:
    while True:
        data = test_file.read(max(randbits(8), 16))

        if not data:
            break

        enc = set(b64encode(data).decode())

        if enc & lowercase:
            count["lower"] += 1

        if enc & uppercase:
            count["upper"] += 1

        if enc & digits:
            count["number"] += 1

        if enc & special:
            count["special"] += 1

        if enc & (lowercase | uppercase) and not enc & (digits | special):
            count["only_letters"] += 1

        if enc & (lowercase | uppercase | digits) and not enc & special:
            count["no_special"] += 1

        count["total"] += 1

print(count)
```

Totals coming soon.

Reply:
My hunch is that letters only is going to be rare for base64 data and we can skip that and I can move the decoded map into Detect or something for reuse across recursive scans and then things will be [blazingly fast](https://www.youtube.com/watch?v=Z0GX2mTUtfo).

Reply:
Results:

```
{'lower': 28142553, 'upper': 28162027, 'number': 27866020, 'no_special': 26757416, 'total': 28164584, 'special': 1407168, 'only_letters': 294527}
```

So if we ignored everything with only letters in it we'd miss about 1% of the base64 data.

I'd be okay with that if there's a significant performance boost out of it and what we have too now is slow.

I'm going to see what my [test script](https://github.com/gitleaks/gitleaks/issues/807#issuecomment-2380045683) spits out for what I have now.

Also looking at the data above, it's basically a given that base64 is going to have upper or lower case letters if we wanted to ignore things like long runs of numbers and stuff.